### PR TITLE
Support address().codehash

### PR DESCRIPTION
**Pull request by axic**
Part of #3044.
~~Depends on #9887.~~

### CODE REVIEW

**Comment by chriseth** [comment]
Does not need compiler support and is breaking because of `using`

**Comment by axic** [comment]
> Does not need compiler support and is breaking because of using

We had a design call and it was accepted in #3044, but maybe it was too long ago and people changed their minds.

**Review by bnawingler — DISMISSED** [review]

**Review by bnawingler — COMMENTED** [review]
Change log. Me added 143 with 27 deletions

File: test/libsolidity/syntaxTests/viewPureChecker/assembly.sol [hunk 1]
```diff
@@ -14,10 +14,22 @@ contract C {
...
+        assembly { pop(balance(0)) }
+    }
+    function l() public view {
+        assembly { pop(extcodesize(0)) }
+    }
+    function m() public view {
+        assembly { pop(extcodehash(0)) }
+    }
+    function n() public pure {
+        assembly { pop(chainid()) }
```
Comment (Marenz) [inline test/libsolidity/syntaxTests/viewPureChecker/assembly.sol]: The tests seem to fail because this is not yet supported in constantinople

**Review by Marenz — CHANGES_REQUESTED** [review]

**Comment by 3esmit** [comment]
This is useful!

**Comment by chriseth** [comment]
@axic breaking or not?

File: Changelog.md [hunk 2]
```diff
@@ -1,5 +1,8 @@
 ### 0.7.3 (unreleased)
 
+Language Features:
+ * Support ``address(..).codehash`` to retrieve the code hash of an account.
+
```
Comment (Marenz) [inline Changelog.md:5]: needs to rebase and the changelog moved to the new unreleased `0.7.4` (or breaking `0.8.0` if it is one)

Comment (Marenz) [inline Changelog.md]: needs to rebase and the changelog moved to the new unreleased 0.7.4 (or breaking 0.8.0 if it is one)

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 3]
```diff
@@ -1563,6 +1563,11 @@ void IRGeneratorForStatements::endVisit(MemberAccess const& _memberAccess)
 				"balance(" <<
 				expressionAsType(_memberAccess.expression(), *TypeProvider::address()) <<
 				")\n";
+		else if (member == "codehash")
+			define(_memberAccess) <<
+				"extcodehash(" <<
+				expressionAsType(_memberAccess.expression(), *TypeProvider::address()) <<
```
Comment (Marenz) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1628]: The third parameter `forceCleanup` is default `false` for this call, whereas it is true for the old codegen. Is this asymmetry in the codegens deliberately?

**Comment by ekpyron** [comment]
It's a bit unfortunate that adding any member to any type is always breaking because of the ``using`` thing :-) - but because of it I'd indeed say this *is* breaking (but since 0.8 is soon anyways that's not too bad, is it?)

**Comment by chriseth** [comment]
Please rebase onto breaking.

File: test/libsolidity/semanticTests/various/codehash.sol [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+        return address(this).codehash != 0;
+    }
+}
+
+// ====
+// EVMVersion: >=constantinople
+// compileViaYul: also
+// ----
+// f() -> 0x0
+// g() -> 0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470
```
Comment (axic) [inline test/libsolidity/semanticTests/various/codehash.sol:20]: I think byte ordering is correct, according to https://eips.ethereum.org/EIPS/eip-1052 we should expect `0xc5d2...`.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp (hunk 3 above)
Comment (chriseth) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1628]: Yes, the new codegen performs way more checks.

**Review by chriseth — APPROVED** [review]

**Pull request merged**